### ISSUE TITLE
Adyen Magento2 Module should follow Google's recommendation for 3rd party POST to mitigate SameSite=Lax cookie changes

### ISSUE THREAD

**Issue by dufrygrant** [Adyen/adyen-magento2#796]
Adyen has a document regarding the SameSite=Lax cookie treatment in recent Chrome versions.

This document links to a Google document about 3D Secure payment flows and recommends a response to this change that requires the 3rd party POST request after 3D Secure verification to operate without the assistance of cookies being set as SameSite=NONE (ie going backwards in terms of security; with Chrome now operating in SameSite=Lax mode by default, this fatally breaks the 3D Secure payment flow, leaving orders in the Payment Review or similar status). This Google-supplied fix is described as "Do not require cookies on the returning POST request".

Adyen's Document:
https://www.adyen.help/hc/en-us/articles/360013000160-How-does-the-Chrome-SameSite-Cookie-policy-affect-my-integration-

Google's Document:
https://github.com/GoogleChromeLabs/samesite-examples/blob/3d-secure-impl/3d-secure.md/

Feature Request:
- The Adyen module's /adyen/redirect/process transaction should be able to operate without PHPSESSID (and other cookies) being set as SameSite=NONE, marking the transaction in the appropriate way as having been completed, followed by a 303 or other similar redirect to the normal Magento /success page, as per the Google suggestion
- This should operate for both logged-in and guest Magento transactions

Reasoning:
- Without such a fix or without the degradation of user security via SameSite=NONE, the Adyen Magento2 module can no longer complete a 3D-secure payment flow successfully with recent Chrome browsers.
- As the Adyen module's controller handles the /adyen/redirect/process request, it should be possible to adjust this without Adyen referring customers to a Magento2 Github thread that simply reduces customer security through setting SameSite=NONE on essential Magento cookies.
- As at 18th August 2020, no adjustment to Magento's own behaviour (such as setting SameSite=NONE on certain cookies) is not yet fixed in any released Magento platform (2.4.x or 2.3.x)

**Comment by david-fuehr** [comment]
We encounter the same issue. If a credit card payment (3ds) can not be completed and the customer has to return to his cart, the complete checkout session is stuck. This is a critical issue, as affected customers:
1. are not able to check out
1. do not know what happend to their order 
1. do not know, how to fix it

Magento support gave us the following explanation and hint:

> The error related to payment caused by the latest updates from the browsers: https://www.chromestatus.com/feature/5088147346030592
> 
> Magento use SameSite = LAX by default. This configuration is recommended for security reasons. SameSite = LAX - is disabling transferring COOKIE to external URLs via POST requests. Magento will not change LAX to None by security reason. You have a problem with your payment integration because the COOKIE where send with post requests.
As a workaround, you can change the SameSite value to None in the webserver configuration. Magento does not recommend such modifications.
> 
> Your payment integration has the wrong architecture and will not work with new security police from Google Chrome, Firefox, and other new versions of browsers.
> 
> Our recommendation is to update payment integration to using The Secure Tokens for authorizations instead of COOKIES.
> 
> Example how Magento did this for PayPal Payflow Pro you can find here https://github.com/magento/magento2/issues/26840#issuecomment-592074494

**Comment by dufrygrant** [comment]
The conclusion I reached was that using nginx to override Set-Cookie headers to SameSite=NONE on certain cookies was the only viable short-term workaround until Adyen alters the operation of the payment module to not require (at a minimum PHPSESSID) on the 3rd party POST call coming back from the 3DS certifier.  That is, to use the first of the two modes of operation that Google recommended in their SameSite/3DS Flow post.

It's viable as a short term fix simply because you can set your nginx to only use the setting for Chrome clients matching a particular user-agent string (and the other browsers aren't as 'tight' on SameSite= for now), but it's not a good long term strategy if all other browsers will be doing the same in due course.  Put another way, it's no worse than a month ago when SameSite was warning but not erroring; as soon as a real fix is available, it should be used in preference to this.

**Comment by david-fuehr** [comment]
FYI: I just found out, that https://github.com/Adyen/adyen-magento2/pull/799 seems to address this issue.

**Comment by acampos1916** [comment]
Hello @david-fuehr @dufrygrant,

Yes, #799 fixes the SameSite cookie issue breaking the user session on 3DS redirect back to the shop. It makes use of Adyen's `redirectFromIssuerMethod` and `redirectToIssuerMethod` to handle the response via GET and not POST. Let us know if you have any feedback.

Thanks,
Ángel

**Comment by fahu** [comment]
We are experiencing the same issue and would like to set `SameSame=none` as a short-term mitigation. Which cookies require this policy?

**Comment by acampos1916** [comment]
Hi @fahu,

Adyen could use the following cookies in the Magento plugin depending on the setup:

__cfduid
_fbp
_ga
_gcl_au
_gid
_hjid
_mkto_trk
gdpr
lastUpdatedGdpr

**Comment by fahu** [comment]
Huge thanks for the information @acampos1916, this helped a lot. (I just realised that I did not thank you when you replied)

**Comment by dufrygrant** [comment]
Hello @acampos1916 I was the original reporter of this back in August.

We're now using v6.6.5.
The site scenario we have is a login-only (Private Sale) Magento site.

When the redirect comes back from the 3DS1 card processor, it's coming back as a POST with essentially identical parameters as what was described in August - the 3rd party post doesn't have an Adyen-set Cookie as far as we can see (there's nothing with a useful or visible name) and when the redirect comes back to the Private Sale site, Magento's redirecting the user back to the account login page, which, now that this is a 1st party GET, has all the cookies and sends the customer to the Account page.

The transaction, which did have 3DS successfully completed on it, now sits in Payment Review and is unrecoverable inside Magento (cannot convert status) and cannot be cancelled prior to the 12hr timeout on Offers inside the Adyen platform.

Now, the behaviour is even worse (at least on a private sale site). The POST is including the correct PHPSESSID, it's including the X-Magento-Vary and even the private_content_version and form_key, but the URL it's hitting (adyen/transparent/redirect/) is getting a 302.

I would note that the earlier GET hit to adyen/process/redirect/ is working just fine; it's just the transparent redirect that's failing.

**Comment by acampos1916** [comment]
Hey @dufrygrant thanks for reaching out with this update.

Is there anything in the logs indicating what could be the problem? Also, maybe you're able to use a debugger in adyen/transparent/redirect/ to understand where the 302 is triggered.

The private sale site should work just as a regular Magento, unless there's a customization or plugin doing extra logic to block controllers only intended for logged in users. If that's the case maybe you can try adding an exception rule for the transparent controller.

Best,
Ángel

**Comment by dufrygrant** [comment]
Hi,

We have implemented an exception rule for the transparent controller (into a webrestrictions.xml file; see below).

What we are now seeing is that when we test 6.5.5 on our UAT site (with Adyen’s 3DS1 test cards, and test username/password 3DS1 service), the POST that comes back does work, even on the ‘private sale’ website.  But, the POST itself seems to differ in a couple of ways from the one coming back from a LIVE 3DS service on our PROD website (where the transaction fails in exactly the same way as it did before; the exception mentioned below doesn’t seem to have any effect).

In short, the Adyen UAT test 3DS1 service works fine, but with the same code inside Magento, same config but with PROD (real, third party) 3DS redirections, it fails.

Grant

<?xml version="1.0"?>

<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_WebsiteRestriction:etc/webrestrictions.xsd">
    <action path="adyen_transparent_redirect" type="generic" />
</config>

**Comment by acampos1916** [comment]
Hello @dufrygrant, could you indicate specifically what are the differences in the POST from test and live? I believe the parameters length might change but the general structure should be the same. Is it possible that your web restrictions file is not taken into account in your live site?

**Comment by dufrygrant** [comment]
Hi,

We tested again today.  It’s working now.  Same files as discussed, no new deployment on our side etc.

It is possible that something changed on the Adyen side for this?

Grant
